src/module_wrap.cc

```diff
@@ -143,38 +140,68 @@ void ModuleWrap::New(const FunctionCallbackInfo<Value>& args) {
     column_offset = Integer::New(isolate, 0);
   }
 
-  ShouldNotAbortOnUncaughtScope no_abort_scope(env);
-  TryCatchScope try_catch(env);
+  Local<String> url;
   Local<Module> module;
 
   Local<PrimitiveArray> host_defined_options =
       PrimitiveArray::New(isolate, HostDefinedOptions::kLength);
   host_defined_options->Set(isolate, HostDefinedOptions::kType,
                             Number::New(isolate, ScriptType::kModule));
 
-  // compile
-  {
-    ScriptOrigin origin(url,
-                        line_offset,                          // line offset
-                        column_offset,                        // column offset
-                        True(isolate),                        // is cross origin
-                        Local<Integer>(),                     // script id
-                        Local<Value>(),                       // source map URL
-                        False(isolate),                       // is opaque (?)
-                        False(isolate),                       // is WASM
-                        True(isolate),                        // is ES Module
-                        host_defined_options);
-    Context::Scope context_scope(context);
-    ScriptCompiler::Source source(source_text, origin);
-    if (!ScriptCompiler::CompileModule(isolate, &source).ToLocal(&module)) {
-      if (try_catch.HasCaught() && !try_catch.HasTerminated()) {
-        CHECK(!try_catch.Message().IsEmpty());
-        CHECK(!try_catch.Exception().IsEmpty());
-        AppendExceptionLine(env, try_catch.Exception(), try_catch.Message(),
-                            ErrorHandlingMode::MODULE_ERROR);
-        try_catch.ReThrow();
+  // new ModuleWrap(syntheticExecutionFunction, export_names, url)
```

Comment:
I won't block on this, but dispatching polymorphism from the js-side (having separate constructors) keeps our native code a lot cleaner.

Reply:
Having a single base-level module object is important for ensuring linking consistency (same linking functions). An architecture where there is `ModuleWrap` which extends into `SourceModuleWrap` and `SyntheticModuleWrap` subclasses could definitely be a good idea, but that should be made in a future refactoring PR.